sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala

```diff
@@ -259,6 +259,55 @@ class V2ExpressionBuilder(
       } else {
         None
       }
+    case date: DateAdd =>
+      val childrenExpressions = date.children.flatMap(generateExpression(_))
+      if (childrenExpressions.length == date.children.length) {
+        Some(new GeneralScalarExpression("DATE_ADD", childrenExpressions.toArray[V2Expression]))
+      } else {
+        None
+      }
+    case date: DateDiff =>
+      val childrenExpressions = date.children.flatMap(generateExpression(_))
+      if (childrenExpressions.length == date.children.length) {
+        Some(new GeneralScalarExpression("DATE_DIFF", childrenExpressions.toArray[V2Expression]))
+      } else {
+        None
+      }
+    case date: TruncDate =>
+      val childrenExpressions = date.children.flatMap(generateExpression(_))
+      if (childrenExpressions.length == date.children.length) {
+        Some(new GeneralScalarExpression("TRUNC", childrenExpressions.toArray[V2Expression]))
+      } else {
+        None
+      }
+    case Second(child, _) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("SECOND", Array[V2Expression](v)))
+    case Minute(child, _) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("MINUTE", Array[V2Expression](v)))
+    case Hour(child, _) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("HOUR", Array[V2Expression](v)))
+    case Month(child) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("MONTH", Array[V2Expression](v)))
+    case Quarter(child) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("QUARTER", Array[V2Expression](v)))
+    case Year(child) => generateExpression(child)
+      .map(v => new GeneralScalarExpression("YEAR", Array[V2Expression](v)))
+    // The DAY_OF_WEEK function in Spark returns the day of the week for date/timestamp.
+    // Database dialects should avoid to follow ISO semantics when handling DAY_OF_WEEK.
+    case DayOfWeek(child) => generateExpression(child)
```

Comment:
It's a bit weird to expose Spark-specific functions to the data source. Spark does support `extract('dayofweek_iso', ...)`, which ends up with `Add(WeekDay(source), Literal(1))`. Shall we match it instead?

Reply:
DayOfWeek is also available in other databases. 
But dayofweek_iso is not common. 
So only the DayOfWeek function is translated, not the WeekDay function.